### PR TITLE
Command Line Interface

### PR DESCRIPTION
**Pull request by vjsrinivas**
As requested, here is the first implementation of #25 with all the core features from the GUI tool.
I've adjusted my `test_cli.py` to utilize some of the tests done in the respective COCO and VOC tests, but I'm open to expanding the tests to make sure everything is truly kosher.

### CODE REVIEW

File: README.md
Comment (rafaelpadilla) [inline README.md:337]: Shouldn't it be ".... there is no visualization functionality for the CLI." ?

File: README.md
Comment (rafaelpadilla) [inline README.md:338]: As other arguments separate the words by "_", shouldn't it be `--gt_format`... or better: '--format_gt` ?

File: README.md
Comment (rafaelpadilla) [inline README.md:339]: As other arguments separate the words by "_", shouldn't it be 'format_det'?

File: README.md
Comment (rafaelpadilla) [inline README.md:340]: another possible option would be `coco`, right?
Shouldn't `voc2012` be in the **Default** column?
Shouldn't `coco` be in the **Notes** column?

File: README.md
Comment (rafaelpadilla) [inline README.md:344]: what is its default path?
shoundn't it be `--save_path`?

File: README.md
Comment (rafaelpadilla) [inline README.md:345]: why not `--coord_det`?

File: cli.py
Comment (rafaelpadilla) [inline cli.py:249]: I'd say "Running metric with VOC2012 metric, using the every point interpolation" instead of "Running metric with VOC2012 metric; AUC (Area Under Curve)"

File: cli.py
Comment (rafaelpadilla) [inline cli.py:232]: I'd include the approach name "Running metric with VOC2012 metric, using the 11-point interpolation approach" instead of simply "Running metric with VOC2007"

File: README.md
Comment (vjsrinivas) [reply README.md:337]: Yup, thank you for catching that typo!

File: README.md
Comment (vjsrinivas) [reply README.md:338]: I agree on the separation of the keywords, and I'm fine with changing both ground truth and detection to `--format_...`

File: README.md
Comment (vjsrinivas) [reply README.md:340]: Yes you're right, there is no options for `--metrics` in the **Notes** column. I'll add all three options. Currently, the CLI requires the user to define a metric; otherwise it will error out. That's why there is no default value in the doc table. If you want there to be a default value, I'm fine with that.

File: README.md
Comment (vjsrinivas) [reply README.md:344]: I guess the example and notes for that argument are confusing... Right now, `-pr` will error out if `--savepath` is not defined or the path does not exist. `--savepath` does not have a default value at the moment.

A possible default value for it could be the current relative path ("./"). How does that sound?

File: cli.py
Comment (vjsrinivas) [reply cli.py]: Fixed in https://github.com/rafaelpadilla/review_object_detection_metrics/pull/57/commits/27394185d27d073b2f871dbd7e836ee038c16ad9

File: cli.py
Comment (vjsrinivas) [reply cli.py]: Fixed in https://github.com/rafaelpadilla/review_object_detection_metrics/pull/57/commits/27394185d27d073b2f871dbd7e836ee038c16ad9

File: README.md
Comment (vjsrinivas) [reply README.md]: Fixed in https://github.com/rafaelpadilla/review_object_detection_metrics/pull/57/commits/27394185d27d073b2f871dbd7e836ee038c16ad9

File: README.md
Comment (vjsrinivas) [reply README.md]: Metric options added in https://github.com/rafaelpadilla/review_object_detection_metrics/pull/57/commits/b7cb0e9e74c1f3f27e90eb83a61db86a3798605c

File: README.md
Comment (vjsrinivas) [reply README.md:339]: Fixed in https://github.com/rafaelpadilla/review_object_detection_metrics/pull/57/commits/b7cb0e9e74c1f3f27e90eb83a61db86a3798605c

File: README.md
Comment (vjsrinivas) [reply README.md]: https://github.com/rafaelpadilla/review_object_detection_metrics/pull/57/commits/b7cb0e9e74c1f3f27e90eb83a61db86a3798605c

File: README.md
Comment (vjsrinivas) [reply README.md:345]: Fixed in https://github.com/rafaelpadilla/review_object_detection_metrics/pull/57/commits/b7cb0e9e74c1f3f27e90eb83a61db86a3798605c

**Comment by vjsrinivas** [comment]
**Question:** Is the `plot_coco_pr_graph` function in `cli.py` redundant? I know there is a PR plotting function for Pascal VOC, but I could not find a similar one for COCO. 

If there isn't one for COCO, I suggest moving the graphing function to `coco_evaluator.py`. If it does exist, please point me to it so I can replace my redundant code.

File: README.md
Comment (rafaelpadilla) [reply README.md:344]: Yes, I think the current relative path ("./") is a good option too. I leave it to you, but we could create a folder in the current relative path to store the files in there, such as ("./results/")

**Comment by rafaelpadilla** [comment]
> **Question:** Is the `plot_coco_pr_graph` function in `cli.py` redundant? I know there is a PR plotting function for Pascal VOC, but I could not find a similar one for COCO.
> 
> If there isn't one for COCO, I suggest moving the graphing function to `coco_evaluator.py`. If it does exist, please point me to it so I can replace my redundant code.

Currently, the AP plotting is made considering only Pascal VOC. While COCO only provides numerical metrics, Pascal VOC provides numerical metrics and the plotting.

As only Pascal VOC official tool provides plottings, for a sake of tools comparison, I think it is better to leave things how they are.

**Comment by vjsrinivas** [comment]
> > **Question:** Is the `plot_coco_pr_graph` function in `cli.py` redundant? I know there is a PR plotting function for Pascal VOC, but I could not find a similar one for COCO.
> > If there isn't one for COCO, I suggest moving the graphing function to `coco_evaluator.py`. If it does exist, please point me to it so I can replace my redundant code.
> 
> Currently, the AP plotting is made considering only Pascal VOC. While COCO only provides numerical metrics, Pascal VOC provides numerical metrics and the plotting.
> 
> As only Pascal VOC official tool provides plottings, for a sake of tools comparison, I think it is better to leave things how they are.

Just clarifying, do you want the PR graphing for COCO disabled all together then? We can throw an exception if the user attempts it.

File: README.md
Comment (vjsrinivas) [reply README.md:344]: Default path "./results/" added for `savepath` in https://github.com/rafaelpadilla/review_object_detection_metrics/pull/57/commits/50d65485d67d6d14e921061c0cf75d8518a648f0

File: cli.py
Comment (rafaelpadilla) [inline cli.py:46]: popular cli tools usually call it `--plot`

File: cli.py
Comment (rafaelpadilla) [inline cli.py:47]: to keep consistency with other arguments, I think it is better to separate words by underline: `--save_path`

File: cli.py
Comment (rafaelpadilla) [inline cli.py:91]: COCO's official tool does provide plots. Only PASCAL VOC does.
So, to avoid making users confused, we better only replicate what the official tools do.

The '--plot' argument should only generate the plot using the `pascal_voc_evaluator.plot_precision_recall_curve` function, if and only if `args.metric=coco`.

**Review by rafaelpadilla — CHANGES_REQUESTED** [review]
I think only these changes are needed before we merge them into the main branch.

By the way, you did an excellent work @vjsrinivas !!!!

**Comment by vjsrinivas** [comment]
Hi @rafaelpadilla, I'm sorry about the lack of response. I got very busy with work, but I'll try to look over your reviews this weekend.

File: cli.py
Comment (vjsrinivas) [reply cli.py]: Fixed in 780b32cd8054d602ec7d9679ae0f4dcc9e4d8943

File: cli.py
Comment (vjsrinivas) [reply cli.py]: Fixed in 780b32cd8054d602ec7d9679ae0f4dcc9e4d8943

File: cli.py
Comment (vjsrinivas) [reply cli.py]: I am assuming you mean that the VOC metric is the only one that can produce the plots? I removed `plot_coco_pr_graph` in 780b32cd8054d602ec7d9679ae0f4dcc9e4d8943

File: README.md
Comment (rafaelpadilla) [inline README.md:345]: The formats for detection will consider <class_id> or <class_name>?
As the parameter `--name` points to the list of names, I suppose only <class_id> are supported, right? We should support the same options as the interface.
What if the user has detections using <class_names>, as in:
`dog 0.78 13 56 100 200`? Is he able to use the CLI?

**Review by rafaelpadilla — CHANGES_REQUESTED** [review]
@vjsrinivas, please, see my comments. I think there are still things needed to be done.

File: cli_test.sh
Comment (vjsrinivas) [reply cli_test.sh]: That was unintentionally committed! 😬 Fixed in 3eadc0de53b2af70b59d6f23f1095078d4bf52bb

File: cli.py
Comment (vjsrinivas) [reply cli.py]: Cleaned up comments in cli.py in 3eadc0de53b2af70b59d6f23f1095078d4bf52bb

File: README.md
Comment (vjsrinivas) [reply README.md]: Removed mention of AUC in 3eadc0de53b2af70b59d6f23f1095078d4bf52bb

File: README.md
Comment (vjsrinivas) [reply README.md:345]: I believe <class_name> is supported in the CLI since it passes `test_case_1`. When specified, `--names` is used for YOLO ground truth (if that gt format is specified) and to replace <class_id> with <class_name>. If `--names` is not specified, CLI will throw error if ground truth is YOLO and assume detection files are in <class_name> format. Unless I'm mistaken, I believe this follows the same behavior as the interface? Are you wanting maybe two different name arguments; one for gt and one for det?

For now, I've updated the old replacement code with the original `replace_id_with_classes` function to reduce clutter (6e1c5b129e50d895a16cd36c321624473ca02940)

I also made an additional `test_case_2`, where the detection files have <class_id> and ground truth files have <class_name> to test if the behavior when `--names` is specified is correct. I won't push the test stuff unless you are interested.

**Comment by vjsrinivas** [comment]
@rafaelpadilla Sorry for the delay! Let me know if there are any other modifications needed.

**Comment by Xu-Justin** [comment]
Still waiting for this to be approved :smile: I need a CLI version to automate my benchmarking system.

**Comment by vjsrinivas** [comment]
@Xu-Justin Unfortunately, I'm very busy with my studies and haven't gotten feedback since the last branch commit.
You can take a look at [my branch](https://github.com/vjsrinivas/review_object_detection_metrics/tree/cli) and take the `cli.py`. I'm not sure if it still works with the latest version of the repo, but it is worth a shot.

**Pull request closed without merging**
